docs/proposals/015-bindless-t.md

```diff
@@ -0,0 +1,160 @@
+This means that you can use a `Bindless<T>` type in any context where an ordinary data type, e.g. `int` type
+is allowed, such as in buffer elements.
+
+On targets where resource handles are already concrete and sized types, `Bindless<T>` simply translates to just `T`.
+
+### Obtaining Actual Resource Handle from `Bindless<T>`
+
+Depending on the target platform and the design choices of the user's application, the way to obtain the actual
+resource handle from a `Bindless<T>` integer handle can vary. Slang does not dictate how this conversion is done,
+and instead, this is left to the user via Slang's link-time specialization ability.
+
+Slang defines the following core module declarations:
+
+```slang
+extern T getResourceFromBindlessHandle(uint64_t handle) where T : IOpaqueHandle
+{
+    // Default Implementation
+    // ...
+}
+```
+
+The `getResourceFromBindlessHandle` is used to convert from a bindless handle to actual opaque resource handle.
+If this function is not provided by the user, the default implementation defined in the core module will be used.
+
+By default, the core module implementation of `getResourceFromBindlessHandle` should use the `ResourceDescriptorHeap` and
+`SamplerDescriptorHeap` builtin object when generating HLSL code. When generating code on other targets, `getResourceFromBindlessHandle`
+will fetch the resource handle from a system defined global array of the corresponding resource type.
+
+If/when SPIRV is extended to expose similar capabilities as D3D's `ResourceDescriptorHeap` feature, we should change the default implementation
+to use that instead. Until we know the default implementation of `getResourceFromBindlessHandle` is stable, we should advise users
+to provide their own implementation of `getResourceFromBindlessHandle` to prevent breakages.
+
+If the user application requires a different bindless implementation, this default behavior can be overrided by defining
+`getResourceFromBindlessHandle` in the user code. Below is a possible user-space implementation of `getResourceFromBindlessHandle`
+for Vulkan:
+
+```slang
+
+// All texture and buffer handles are defined in descriptor set 100.
+[vk::binding(0, 100)]
+__DynamicResource<__DynamicResourceKind.General> resourceHandles[];
+
+// All sampler handles are defined in descriptor set 101.
+[vk::binding(0, 101)]
+__DynamicResource<__DynamicResourceKind.Sampler> samplerHandles[];
+
+export getResourceFromBindlessHandle<T>(uint64_t handle) where T : IOpaqueHandle
+{
+    if (T.getResourceKind() == ResourceKind.Sampler)
+        return (T)samplerHandles[handle];
+    else
+        return (T)resourceHandles[handle];
+}
+```
+
+### Invalid Handle
+
+We reserve `uint64_t.maxValue` as a special handle value of `Bindless<T>` types to mean an invalid/null resource.
+This will allow us to optimize `Optional<Bindless<Texture2D>>` to use the reserved value to mean no-value.
```

Comment:
I initially didn't catch this sentence, which seems to be very important.

Notably, we seem to be saying that in all other contexts, the Slang compiler will assume that a non-optional `Bindless<T>` has a valid handle value, and that it is programmer error (I'm going to hold back from saying "undefined behavior") if they attempt to use a `Bindless<T>` that has an invalid handle value.

(This isn't actually surprising, since there will no doubt be many integers that don't represent valid handles, and using any of those would be a programmer error too. The point is more that the compiler is allowed to assume a `Bindless<T>` is valid, and programmers are expected to use an `Optional<>` if they want to allow for invalid handles.)

Reply:
I assume that `Optional<Bindless<T>>` will Just Work with an `if let`?

Reply:
Yes, it should.